File: ms/controllerblueprints/modules/service/src/main/java/org/onap/ccsdk/apps/controllerblueprints/service/BluePrintEnhancerService.java
/*
 * Copyright © 2017-2018 AT&T Intellectual Property.
 * Modifications Copyright © 2018 IBM.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.onap.ccsdk.apps.controllerblueprints.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.base.Preconditions;
import org.apache.commons.collections.MapUtils;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.onap.ccsdk.apps.controllerblueprints.core.BluePrintException;
import org.onap.ccsdk.apps.controllerblueprints.core.ConfigModelConstant;
import org.onap.ccsdk.apps.controllerblueprints.core.data.*;
import org.onap.ccsdk.apps.controllerblueprints.core.service.BluePrintEnhancerDefaultService;
import org.onap.ccsdk.apps.controllerblueprints.core.service.BluePrintRepoService;
import org.onap.ccsdk.apps.controllerblueprints.core.utils.JacksonUtils;
import org.onap.ccsdk.apps.controllerblueprints.resource.dict.ResourceAssignment;
import com.att.eelf.configuration.EELFLogger;
import com.att.eelf.configuration.EELFManager;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * BluePrintEnhancerService
 *
 * @author Brinda Santh DATE : 8/8/2018
 */

@Service
public class BluePrintEnhancerService extends BluePrintEnhancerDefaultService {

    private static EELFLogger log = EELFManager.getInstance().getLogger(BluePrintEnhancerService.class);

    private Map<String, DataType> recipeDataTypes = new HashMap<>();

    public BluePrintEnhancerService(BluePrintRepoService bluePrintEnhancerRepoDBService) {
        super(bluePrintEnhancerRepoDBService);
    }

    @Override
    public void enrichTopologyTemplate(@NotNull ServiceTemplate serviceTemplate) throws BluePrintException{
        super.enrichTopologyTemplate(serviceTemplate);

        // Update the Recipe Inputs and DataTypes
        populateRecipeInputs(serviceTemplate);
    }


    @Override
    public void enrichNodeTemplate(@NotNull String nodeTemplateName, @NotNull NodeTemplate nodeTemplate) throws BluePrintException {
        super.enrichNodeTemplate(nodeTemplateName, nodeTemplate);

        String nodeTypeName = nodeTemplate.getType();
        log.info("*** Enriching NodeType: {}", nodeTypeName);
        // Get NodeType from Repo and Update Service Template
        NodeType nodeType = super.populateNodeType(nodeTypeName);

        // Enrich NodeType
        super.enrichNodeType(nodeTypeName, nodeType);

        // Custom for Artifact Population
        if (StringUtils.isNotBlank(nodeType.getDerivedFrom())
                && ConfigModelConstant.MODEL_TYPE_NODE_ARTIFACT.equalsIgnoreCase(nodeType.getDerivedFrom())) {
            populateArtifactTemplateMappingDataType(nodeTemplateName, nodeTemplate);
        }

        //Enrich Node Template Artifacts
        super.enrichNodeTemplateArtifactDefinition(nodeTemplateName, nodeTemplate);

    }


    private void populateArtifactTemplateMappingDataType(@NotNull String nodeTemplateName, @NotNull NodeTemplate nodeTemplate)
            throws BluePrintException {
        log.info("****** Processing Artifact Node Template : {}", nodeTemplateName);

        if (nodeTemplate.getProperties() != null) {

            if (!nodeTemplate.getProperties().containsKey(ConfigModelConstant.PROPERTY_RECIPE_NAMES)) {
                throw new BluePrintException("Node Template (" + nodeTemplateName + ") doesn't have "
                        + ConfigModelConstant.PROPERTY_RECIPE_NAMES + " property.");
            }

            // Modified for ONAP converted Object to JsonNode
            JsonNode recipeNames = nodeTemplate.getProperties().get(ConfigModelConstant.PROPERTY_RECIPE_NAMES);

            log.info("Processing Receipe Names : {} ", recipeNames);

            if (recipeNames != null && recipeNames.isArray() && recipeNames.size() > 0) {

                Map<String, PropertyDefinition> mappingProperties =
                        getCapabilityMappingProperties(nodeTemplateName, nodeTemplate);

                for (JsonNode recipeNameNode : recipeNames) {
                    String recipeName = recipeNameNode.textValue();
                    processRecipe(nodeTemplateName, mappingProperties, recipeName);
                }
            }
        }
    }

    private void processRecipe(@NotNull String nodeTemplateName, Map<String, PropertyDefinition> mappingProperties, String recipeName) {
        if (StringUtils.isNotBlank(recipeName)) {
            DataType recipeDataType = this.recipeDataTypes.get(recipeName);
            if (recipeDataType == null) {
                log.info("DataType not present for the recipe({})", recipeName);
                recipeDataType = new DataType();
                recipeDataType.setVersion("1.0.0");
                recipeDataType.setDescription(
                        "This is Dynamic Data type definition generated from resource mapping for the config template name "
                                + nodeTemplateName + ".");
                recipeDataType.setDerivedFrom(ConfigModelConstant.MODEL_TYPE_DATA_TYPE_DYNAMIC);
                Map<String, PropertyDefinition> dataTypeProperties = new HashMap<>();
                recipeDataType.setProperties(dataTypeProperties);
            } else {
                log.info("DataType Already present for the recipe({})", recipeName);
            }

            // Merge all the Recipe Properties
            mergeDataTypeProperties(recipeDataType, mappingProperties);

            // Overwrite Recipe DataType
            this.recipeDataTypes.put(recipeName, recipeDataType);

        }
    }

    private Map<String, PropertyDefinition> getCapabilityMappingProperties(String nodeTemplateName,
                                                                           NodeTemplate nodeTemplate) throws BluePrintException {

        Map<String, PropertyDefinition> dataTypeProperties = null;
        if (nodeTemplate != null && MapUtils.isNotEmpty(nodeTemplate.getCapabilities())) {
            CapabilityAssignment capability =
                    nodeTemplate.getCapabilities().get(ConfigModelConstant.CAPABILITY_PROPERTY_MAPPING);

            if (capability != null && capability.getProperties() != null) {

                String resourceAssignmentContent = JacksonUtils
                        .getJson(capability.getProperties().get(ConfigModelConstant.CAPABILITY_PROPERTY_MAPPING));

                List<ResourceAssignment> resourceAssignments =
                        JacksonUtils.getListFromJson(resourceAssignmentContent, ResourceAssignment.class);

                Preconditions.checkNotNull(resourceAssignments, "Failed to Processing Resource Mapping " + resourceAssignmentContent);
                dataTypeProperties = new HashMap<>();

                for (ResourceAssignment resourceAssignment : resourceAssignments) {
                    if (resourceAssignment != null
                            // && Boolean.valueOf(resourceAssignment.getInputParameter())
                            && resourceAssignment.getProperty() != null
                            && StringUtils.isNotBlank(resourceAssignment.getName())) {

                        // Enrich the Property Definition
                        super.enrichPropertyDefinition(resourceAssignment.getName(), resourceAssignment.getProperty());

                        dataTypeProperties.put(resourceAssignment.getName(), resourceAssignment.getProperty());

                    }
                }

            }
        }
        return dataTypeProperties;
    }

    private void mergeDataTypeProperties(DataType dataType, Map<String, PropertyDefinition> mergeProperties) {
        if (dataType != null && dataType.getProperties() != null && mergeProperties != null) {
            // Add the Other Template Properties
            mergeProperties.forEach((mappingKey, propertyDefinition) -> dataType.getProperties().put(mappingKey, propertyDefinition));
        }
    }

    private void populateRecipeInputs(ServiceTemplate serviceTemplate) {
        if (serviceTemplate.getTopologyTemplate() != null
                && MapUtils.isNotEmpty(serviceTemplate.getTopologyTemplate().getInputs())
                && MapUtils.isNotEmpty(this.recipeDataTypes)
                && MapUtils.isNotEmpty(serviceTemplate.getDataTypes())) {
            this.recipeDataTypes.forEach((recipeName, recipeDataType) -> {
                String dataTypePrefix = recipeName.replace("-action", "") + "-request";
                String dataTypeName = "dt-" + dataTypePrefix;

                serviceTemplate.getDataTypes().put(dataTypeName, recipeDataType);

                PropertyDefinition customInputProperty = new PropertyDefinition();
                customInputProperty.setDescription("This is Dynamic Data type for the receipe " + recipeName + ".");
                customInputProperty.setRequired(Boolean.FALSE);
                customInputProperty.setType(dataTypeName);
                serviceTemplate.getTopologyTemplate().getInputs().put(dataTypePrefix, customInputProperty);

            });
        }
    }
}

File: ms/controllerblueprints/modules/core/src/main/kotlin/org/onap/ccsdk/apps/controllerblueprints/core/service/BluePrintEnhancerService.kt
/*
 * Copyright © 2017-2018 AT&T Intellectual Property.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.onap.ccsdk.apps.controllerblueprints.core.service

import org.onap.ccsdk.apps.controllerblueprints.core.BluePrintException
import org.onap.ccsdk.apps.controllerblueprints.core.BluePrintTypes
import org.onap.ccsdk.apps.controllerblueprints.core.data.*
import org.onap.ccsdk.apps.controllerblueprints.core.format
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.Serializable

/**
 * BluePrintEnhancerService
 * @author Brinda Santh
 *
 */
interface BluePrintEnhancerService : Serializable {

    @Throws(BluePrintException::class)
    fun enhance(content: String): ServiceTemplate

    @Throws(BluePrintException::class)
    fun enhance(serviceTemplate: ServiceTemplate): ServiceTemplate

    /**
     * Read Blueprint from CSAR structure Directory
     */
    @Throws(BluePrintException::class)
    fun enhance(fileName: String, basePath: String): ServiceTemplate
}

open class BluePrintEnhancerDefaultService(val bluePrintEnhancerRepoService: BluePrintEnhancerRepoService) : BluePrintEnhancerService {

    private val log: Logger = LoggerFactory.getLogger(BluePrintEnhancerDefaultService::class.java)

    lateinit var serviceTemplate: ServiceTemplate

    override fun enhance(content: String): ServiceTemplate {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun enhance(fileName: String, basePath: String): ServiceTemplate {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun enhance(serviceTemplate: ServiceTemplate): ServiceTemplate {
        this.serviceTemplate = serviceTemplate
        initialCleanUp()
        enrichTopologyTemplate(serviceTemplate)

       // log.info("Enriched Blueprint :\n {}", JacksonUtils.getJson(serviceTemplate, true))
        return this.serviceTemplate
    }

    open fun initialCleanUp() {
        serviceTemplate.artifactTypes?.clear()
        serviceTemplate.nodeTypes?.clear()
        serviceTemplate.dataTypes?.clear()

        serviceTemplate.artifactTypes = HashMap()
        serviceTemplate.nodeTypes = HashMap()
        serviceTemplate.dataTypes = HashMap()

    }

    open fun enrichTopologyTemplate(serviceTemplate: ServiceTemplate) {
        serviceTemplate.topologyTemplate?.let { topologyTemplate ->
            enrichTopologyTemplateInputs(topologyTemplate)
            enrichTopologyTemplateNodeTemplates(topologyTemplate)
        }
    }

    open fun enrichTopologyTemplateInputs(topologyTemplate: TopologyTemplate) {
        topologyTemplate.inputs?.let { inputs ->
            enrichPropertyDefinitions(inputs)
        }
    }

    open fun enrichTopologyTemplateNodeTemplates(topologyTemplate: TopologyTemplate) {
        topologyTemplate.nodeTemplates?.forEach { nodeTemplateName, nodeTemplate ->
            enrichNodeTemplate(nodeTemplateName, nodeTemplate)
        }
    }

    @Throws(BluePrintException::class)
    open fun enrichNodeTemplate(nodeTemplateName: String, nodeTemplate: NodeTemplate) {
        val nodeTypeName = nodeTemplate.type
        // Get NodeType from Repo and Update Service Template
        val nodeType = populateNodeType(nodeTypeName)

        // Enrich NodeType
        enrichNodeType(nodeTypeName, nodeType)

        //Enrich Node Template Artifacts
        enrichNodeTemplateArtifactDefinition(nodeTemplateName, nodeTemplate)
    }

    open fun enrichNodeType(nodeTypeName: String, nodeType: NodeType) {

        // NodeType Property Definitions
        enrichNodeTypeProperties(nodeTypeName, nodeType)

        //NodeType Requirement
        enrichNodeTypeRequirements(nodeTypeName, nodeType)

        //NodeType Capability
        enrichNodeTypeCapabilityProperties(nodeTypeName, nodeType)

        //NodeType Interface
        enrichNodeTypeInterfaces(nodeTypeName, nodeType)
    }

    open fun enrichNodeTypeProperties(nodeTypeName: String, nodeType: NodeType) {
        nodeType.properties?.let { enrichPropertyDefinitions(nodeType.properties!!) }
    }

    open fun enrichNodeTypeRequirements(nodeTypeName: String, nodeType: NodeType) {

        nodeType.requirements?.forEach { requirementDefinitionName, requirementDefinition ->
            // Populate Requirement Node
            requirementDefinition.node?.let { requirementNodeTypeName ->
                // Get Requirement NodeType from Repo and Update Service Template
                val requirementNodeType = populateNodeType(requirementNodeTypeName)

                enrichNodeType(requirementNodeTypeName, requirementNodeType)
            }
        }
    }

    open fun enrichNodeTypeCapabilityProperties(nodeTypeName: String, nodeType: NodeType) {
        nodeType.capabilities?.forEach { capabilityDefinitionName, capabilityDefinition ->
            capabilityDefinition.properties?.let { properties ->
                enrichPropertyDefinitions(properties)
            }
        }
    }

    open fun enrichNodeTypeInterfaces(nodeTypeName: String, nodeType: NodeType) {
        nodeType.interfaces?.forEach { interfaceName, interfaceObj ->
            // Populate Node type Interface Operation
            log.info("*** ** Enriching NodeType: {} Interface {}", nodeTypeName, interfaceName)
            populateNodeTypeInterfaceOperation(nodeTypeName, interfaceName, interfaceObj)

        }
    }

    open fun populateNodeTypeInterfaceOperation(nodeTypeName: String, interfaceName: String, interfaceObj: InterfaceDefinition) {

        interfaceObj.operations?.forEach { operationName, operation ->
            enrichNodeTypeInterfaceOperationInputs(nodeTypeName, operationName, operation)
            enrichNodeTypeInterfaceOperationOputputs(nodeTypeName, operationName, operation)
        }
    }

    open fun enrichNodeTypeInterfaceOperationInputs(nodeTypeName: String, operationName: String, operation: OperationDefinition) {
        operation.inputs?.let { inputs ->
            enrichPropertyDefinitions(inputs)
        }
    }

    open fun enrichNodeTypeInterfaceOperationOputputs(nodeTypeName: String, operationName: String, operation: OperationDefinition) {
        operation.outputs?.let { inputs ->
            enrichPropertyDefinitions(inputs)
        }
    }

    open fun enrichPropertyDefinitions(properties: MutableMap<String, PropertyDefinition>) {

        properties.forEach { propertyName, propertyDefinition ->
            enrichPropertyDefinition(propertyName, propertyDefinition)
        }
    }

    open fun enrichPropertyDefinition(propertyName: String, propertyDefinition: PropertyDefinition) {
        val propertyType = propertyDefinition.type
                ?: throw BluePrintException(format("Property type is missing for property : {}", propertyName))
        if (BluePrintTypes.validPrimitiveTypes().contains(propertyType)) {

        } else if (BluePrintTypes.validCollectionTypes().contains(propertyType)) {
            val entrySchema = propertyDefinition.entrySchema
                    ?: throw BluePrintException(format("Entry Schema is missing for collection property : {}", propertyName))

            if (!BluePrintTypes.validPrimitiveTypes().contains(entrySchema.type)) {
                populateDataTypes(entrySchema.type)
            }
        } else {
            populateDataTypes(propertyType)
        }

    }

    open fun enrichNodeTemplateArtifactDefinition(nodeTemplateName: String, nodeTemplate: NodeTemplate) {

        nodeTemplate.artifacts?.forEach { artifactDefinitionName, artifactDefinition ->
            val artifactTypeName = artifactDefinition.type
                    ?: throw BluePrintException(format("Artifact type is missing for NodeTemplate({}) artifact({})", nodeTemplateName, artifactDefinitionName))

            // Populate Artifact Type
            populateArtifactType(artifactTypeName)
        }
    }

    open fun populateNodeType(nodeTypeName: String): NodeType {
        val nodeType = bluePrintEnhancerRepoService.getNodeType(nodeTypeName)
                ?: throw BluePrintException(format("Couldn't get NodeType({}) from repo.", nodeTypeName))
        serviceTemplate.nodeTypes?.put(nodeTypeName, nodeType)
        return nodeType
    }

    open fun populateArtifactType(artifactTypeName: String): ArtifactType {
        val artifactType = bluePrintEnhancerRepoService.getArtifactType(artifactTypeName)
                ?: throw BluePrintException(format("Couldn't get ArtifactType({}) from repo.", artifactTypeName))
        serviceTemplate.artifactTypes?.put(artifactTypeName, artifactType)
        return artifactType
    }

    open fun populateDataTypes(dataTypeName: String): DataType {
        val dataType = bluePrintEnhancerRepoService.getDataType(dataTypeName)
                ?: throw BluePrintException(format("Couldn't get DataType({}) from repo.", dataTypeName))
        serviceTemplate.dataTypes?.put(dataTypeName, dataType)
        return dataType
    }

}

